src/torrent.cpp

```diff
@@ -5790,6 +5808,50 @@ namespace {
 		set_error(ec, torrent_status::error_file_none);
 	}
 
+#if TORRENT_USE_RTC
+	void torrent::generate_rtc_offers(int count
+			, std::function<void(error_code const&, std::vector<aux::rtc_offer>)> handler)
+	{
+		// rtc_signaling is created lazily
+		if(!m_rtc_signaling)
+		{
+			m_rtc_signaling = std::make_shared<aux::rtc_signaling>(m_ses.get_context()
+				, this
+				, std::bind(&torrent::on_rtc_stream, this, _1, _2));
+		}
+
+		m_rtc_signaling->generate_offers(count, std::move(handler));
+	}
+
+	void torrent::on_rtc_offer(aux::rtc_offer const& offer)
+	{
+		TORRENT_ASSERT(m_rtc_signaling);
+		if(m_rtc_signaling) m_rtc_signaling->process_offer(offer);
+	}
+
+	void torrent::on_rtc_answer(aux::rtc_answer const& answer)
+	{
+		TORRENT_ASSERT(m_rtc_signaling);
+		if(m_rtc_signaling) m_rtc_signaling->process_answer(answer);
+	}
+
+	void torrent::on_rtc_stream(peer_id const& pid, aux::rtc_stream_init stream_init)
+	{
+		aux::socket_type s(aux::rtc_stream(m_ses.get_context(), stream_init));
+
+		need_peer_list();
+		torrent_state st = get_peer_list_state();
+		torrent_peer* peerinfo = m_peer_list->add_rtc_peer(pid.to_string(), peer_source_flags_t{}, {}, &st);
+
+		error_code ec;
+		auto endpoint = s.remote_endpoint(ec);
+		TORRENT_ASSERT(!ec);
+		if(ec) return;
+
+		create_peer_connection(peerinfo, std::move(s), std::move(endpoint));
```

Comment:
shouldn't you check the return value here? and whether `peerinfo->connection` was set?

Reply:
Yes but what should you do if it returns `false` meaning the peer is actually disconnecting?

Reply:
I guess at that point `s` has been destroyed already anyway